azkaban-common/src/main/java/azkaban/executor/container/ContainerImplUtils.java

```diff
@@ -82,4 +96,103 @@ public static int getFlowNameHashValMapping(ExecutableFlow flow) {
     int flowNameHashVal = Math.abs(MurmurHash3.hash32x86(flowNameBytes));
     return flowNameHashVal % 100 + 1;
   }
-}
+
+  public static Set<String> getProxyUsersForFlow(final ProjectManager projectManager,
+      final ExecutableFlow flow) {
+    final Set<String> proxyUsers = new HashSet<>();
+    // Get the project and flow Object that needs to be used repeatedly in the DAG.
+    Project project = projectManager.getProject(flow.getProjectId());
+    Flow flowObj = project.getFlow(flow.getFlowId());
+    Optional<Props> flowProps;
+
+      /* Get the flow properties and check if the proxy user is present in the highest level of the
+     flow and not the job. Passing null as the job name is able to get us the top level flow
+     properties mentioned for the flow. Since there are overridden in most cases, it's usually
+     not required. We also append the .flow extension as flow's propSource has not be resolved
+     here, which does include the suffix.  */
+
+    if(FlowLoaderUtils.isAzkabanFlowVersion20(flow.getAzkabanFlowVersion())) {
+      flowProps = Optional.ofNullable(projectManager.getProperties(project, flowObj, null,
+          flow.getFlowId() + FLOW_FILE_SUFFIX));
+      flowProps.ifPresent(s -> proxyUsers.add(s.getString(USER_TO_PROXY, "")));
+    } else {
+      // Handle Flow 1.0 Properties fetch.
+      final Map<String, String> flow10Props = new HashMap<>();
+      for (String propsFile : flowObj.getAllFlowProps().keySet()) {
+        flow10Props
+            .putAll(projectManager.getProperties(project, flowObj, null, propsFile).getFlattened());
+      }
+      proxyUsers.add(flow10Props.getOrDefault(USER_TO_PROXY, ""));
+    }
```

Comment:
We might need to revisit this logic in the future depending on how many times `getProxyUsersForFlow` gets called and how long it takes to run for now it should be ok. Issues with it are:
-to work flow id must match the name of the flow file for Flow 2.0.
-DB may be called multiple times to fetch Flow 1.0 props.

Reply:
ack. added TODO.